qiskit/primitives/base_estimator.py

```diff
@@ -0,0 +1,199 @@
+    H2 = SparsePauliOp.from_list([("IZ", 1)])
+    H3 = SparsePauliOp.from_list([("ZI", 1), ("ZZ", 1)])
+
+    with Estimator([psi1, psi2], [H1, H2, H3], list(params1) + list(params2)) as e:
+        # first 6 values correspond to the parameters of psi1
+        # last 8 values correspond to the parameters of psi2
+        theta1 = [0, 1, 1, 2, 3, 5] + [0] * 8
+        theta2 = [0] * 6 + [0, 1, 1, 2, 3, 5, 8, 13]
+        theta3 = [1, 2, 3, 4, 5, 6] + [0] * 8
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)> ]
+        psi1_H1_result = e(theta1, shots=1024, grouping=[(0, 0)])
+        print(psi1_H1_result)
+
+        # calculate [ <psi1(theta1)|H2|psi1(theta1)>, <psi1(theta1)|H3|psi1(theta1)> ]
+        psi1_H23_result = e(theta1, shots=1024, grouping=[(0, 1), (0, 2)])
+        print(psi1_H23_result)
+
+        # calculate [ <psi2(theta2)|H2|psi2(theta2)> ]
+        psi2_H2_result = e(theta2, shots=1024, grouping=[(1, 1)])
+        print(psi2_H2_result)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi1(theta3)|H1|psi1(theta3)> ]
+        psi1_H1_result2 = e([theta1, theta3], shots=1024, grouping=[(0, 0)])
+        print(psi1_H1_result2)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi2(theta2)|H2|psi2(theta2)>, <psi1(theta3)|H3|psi1(theta3)> ]
+        psi12_H23_result = e([theta1, theta2, theta3], shots=1024, grouping=[(0, 0), (1, 1), (0, 2)])
+        print(psi12_H23_result)
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Optional, Union
+
+from qiskit.circuit import QuantumCircuit, Parameter
+from qiskit.circuit.parametertable import ParameterView
+from qiskit.quantum_info import SparsePauliOp
+
+from .estimator_result import EstimatorResult
+
+
+@dataclass(frozen=True)
+class Group:
+    """The dataclass represents indices of circuit and observable."""
+
+    circuit_index: int
+    observable_index: int
+
+
+class BaseEstimator(ABC):
+    """Estimator base class.
+
+    Base class for Estimator that estimates expectation values of quantum circuits and observables.
+    """
+
+    def __init__(
+        self,
+        circuits: list[QuantumCircuit],
+        observables: list[SparsePauliOp],
+        parameters: Union[ParameterView, list[Parameter]],
+    ):
+        """
+        Args:
+            circuits (list[QuantumCircuit]): quantum circuits that represents quantum states
+            observables (list[SparsePauliOp]): observables
+            parameters (Union[ParameterView, list[Parameter]]): parameters of quantum circuits
+        """
+        self._circuits = circuits
+        self._observables = observables
+        self._parameters = parameters
+
+    @abstractmethod
+    def __enter__(self):
+        ...
+
+    @abstractmethod
+    def __exit__(self, ex_type, ex_value, trace):
+        ...
+
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        cls.__call__ = cls.run
```

Comment:
This way of runtime-patching the class def is cute, but I think it messes up IDE intellisense completion and type hinting stuff when calling an instance, so I rather have something like
```suggestion
    def __call__(<signature copied from run()>):
        return self.run(parameters, grouping, **run_options)
```
which at least in VS code enables those features